### PR TITLE
RegisterNonce for waiting and view function 

### PR DESCRIPTION
**Pull request by sasurobert**
RegisterNonce for waiting and view function.
This code is backward compatible if stake was not enabled - after that it is not. Should be merged before enabling staking.

### CODE REVIEW

File: vm/systemSmartContracts/staking.go [hunk 1]
```diff
@@ -1297,6 +1294,52 @@ func (r *stakingSC) getRemainingUnbondPeriod(args *vmcommon.ContractCallInput) v
...
+	nextKey := make([]byte, len(waitingListHead.FirstKey))
+	copy(nextKey, waitingListHead.FirstKey)
+	for len(nextKey) != 0 && index <= waitingListHead.Length {
+		element, err := r.getWaitingListElement(nextKey)
+		if err != nil {
+			r.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		stakedData, err := r.getOrCreateRegisteredData(element.BLSPublicKey)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1327]: it is ok. it shouldn't happen to create here - only bad programming

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

File: process/scToProtocol/stakingToPeer.go [hunk 2]
```diff
@@ -125,13 +117,44 @@ func (stp *stakingToPeer) getPeerAccount(key []byte) (state.PeerAccountHandler,
...
+		return nil, process.ErrWrongTypeAssertion
+	}
+
+	return userAcc, nil
+}
+
+func (stp *stakingToPeer) getStorageFromAccount(userAcc state.UserAccountHandler, key []byte) []byte {
+	value, err := userAcc.DataTrieTracker().RetrieveValue(key)
+	if err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:137]: it is log traced in retrievevalue.

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -1159,12 +1156,12 @@ func (r *stakingSC) getWaitingListIndex(args *vmcommon.ContractCallInput) vmcomm
 		return vmcommon.UserError
 	}
 
-	index := 2
+	index := uint32(2)
 	nextKey := make([]byte, len(waitingElementKey))
 	copy(nextKey, prevElement.NextKey)
-	for len(nextKey) != 0 {
+	for len(nextKey) != 0 && index <= waitingListHead.Length {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1162]: index < waitingListHead.Length ?

File: vm/systemSmartContracts/staking.go [hunk 4]
```diff
@@ -1297,6 +1294,52 @@ func (r *stakingSC) getRemainingUnbondPeriod(args *vmcommon.ContractCallInput) v
...
+	}
+	if waitingListHead.Length == 0 {
+		r.eei.AddReturnMessage("no one in waitingList")
+		return vmcommon.UserError
+	}
+
+	index := uint32(1)
+	nextKey := make([]byte, len(waitingListHead.FirstKey))
+	copy(nextKey, waitingListHead.FirstKey)
+	for len(nextKey) != 0 && index <= waitingListHead.Length {
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1320]: index < waitingListHead.Length ?

File: vm/systemSmartContracts/staking.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1162]: Index starts from 1

File: vm/systemSmartContracts/staking.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1320]: Index starts from 1

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/staking.go [hunk 5]
```diff
@@ -1255,7 +1255,7 @@ func (r *stakingSC) getBLSKeyStatus(args *vmcommon.ContractCallInput) vmcommon.R
 		return vmcommon.Ok
 	}
 	if stakedData.Waiting {
-		r.eei.Finish([]byte("waiting"))
+		r.eei.Finish([]byte("queued"))
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1258]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System test passed.

**Pull request merged**